sparkplug/config/connection.py

```diff
@@ -61,41 +61,49 @@ def locked_fn(*args, **kwargs):
         with lock:
             r = fn(*args, **kwargs)
         return r
-
     return locked_fn
 
 
 class MultiThreadedConnection(object):
     """
     Context Manager
 
-    Replaces the frame_writer on a connection
-    for the purposes of synchronizing use of the connection
-    between threads.
-
-    Also makes connection.transport _read() and _write() thread safe.
+    Replaces methods on connection, channel
+    with ones that use a shared lock; to
+    prevent the consumer and the heartbeater
+    from stepping on each other between threads.
     """
 
-    def __init__(self, connection):
+    def __init__(self, connection, channel):
         self._connection = connection
-        self._holds = {}
+        self._channel = channel
+        self._connection_hold = {}
+        self._channel_hold = {}
         self._lock = threading.RLock()
 
+    def _lock_obj(self, obj, store):
+        elements = set(dir(obj)) & set(['send_heartbeat', 'send_method'])
+        for e in elements:
+            attr = getattr(obj, e)
+            if callable(attr):
+                store[e] = attr
+                setattr(obj, e, _locked_call(self._lock, attr))
```

Comment:
This looks like it can be simplified with [operator.attrgetter](https://docs.python.org/3/library/operator.html#operator.attrgetter):

```suggestion
        function_getter = operator.attrgetter('send_heartbeat', 'send_method')
        for func in filter(callable, function_getter(obj)):
            store[func.__name__] = attr
            setattr(obj, func.__name__, _locked_call(self._lock, func))
```

Reply:
We currently require py 3.5 or better, so operator.attrgetter is an option.